初始化壳子类的时候无法找到

运行时总是报错：
```
2021-04-14 21:38:54.101 10121-10121/com.example.zhanglei.myapplication.test E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.example.zhanglei.myapplication.test, PID: 10121
    java.lang.RuntimeException: Unable to instantiate activity ComponentInfo{com.example.zhanglei.myapplication.test/com.hl.my_runtime.PluginDefaultProxyActivity}: java.lang.ClassNotFoundException: Didn't find class "com.hl.my_runtime.PluginDefaultProxyActivity" on path: DexPathList[[zip file "/data/app/com.example.zhanglei.myapplication.test-gqeKhx0oVxhB_YnymbuOOQ==/base.apk"],nativeLibraryDirectories=[/data/app/com.example.zhanglei.myapplication.test-gqeKhx0oVxhB_YnymbuOOQ==/lib/arm64, /data/app/com.example.zhanglei.myapplication.test-gqeKhx0oVxhB_YnymbuOOQ==/base.apk!/lib/arm64-v8a, /system/lib64, /hw_product/lib64, /system/product/lib64, /prets/lib64]]
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3812)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4077)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:91)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:149)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:103)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2458)
        at android.os.Handler.dispatchMessage(Handler.java:110)
        at android.os.Looper.loop(Looper.java:219)
        at android.app.ActivityThread.main(ActivityThread.java:8387)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1055)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "com.hl.my_runtime.PluginDefaultProxyActivity" on path: DexPathList[[zip file "/data/app/com.example.zhanglei.myapplication.test-gqeKhx0oVxhB_YnymbuOOQ==/base.apk"],nativeLibraryDirectories=[/data/app/com.example.zhanglei.myapplication.test-gqeKhx0oVxhB_YnymbuOOQ==/lib/arm64, /data/app/com.example.zhanglei.myapplication.test-gqeKhx0oVxhB_YnymbuOOQ==/base.apk!/lib/arm64-v8a, /system/lib64, /hw_product/lib64, /system/product/lib64, /prets/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:209)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at android.app.AppComponentFactory.instantiateActivity(AppComponentFactory.java:95)
        at androidx.core.app.CoreComponentFactory.instantiateActivity(CoreComponentFactory.java:45)
        at android.app.Instrumentation.newActivity(Instrumentation.java:1264)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3797)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4077) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:91) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:149) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:103) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2458) 
        at android.os.Handler.dispatchMessage(Handler.java:110) 
        at android.os.Looper.loop(Looper.java:219) 
        at android.app.ActivityThread.main(ActivityThread.java:8387) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1055) 
```

宿主中定义的壳子类如下：
```xml
        <activity
            android:name="com.hl.my_runtime.PluginDefaultProxyActivity"
            android:launchMode="standard"
            android:screenOrientation="portrait"
            android:configChanges="mcc|mnc|locale|touchscreen|keyboard|keyboardHidden|navigation|screenLayout|fontScale|uiMode|orientation|screenSize|smallestScreenSize|layoutDirection"
            android:hardwareAccelerated="true"
            android:theme="@style/PluginContainerActivity"
            android:multiprocess="true"
            android:process=":plugin"
            />
```

壳子类定义如下图：
![image](https://user-images.githubusercontent.com/33248551/114720922-0243ee80-9d6b-11eb-904b-e5526334c6ac.png)

插件的 applicationId 已与宿主保持一致，不知道为啥会出现此问题，希望可以指点迷津，提前谢谢！

--- a/projects/sdk/core/loader/src/main/kotlin/com/tencent/shadow/core/loader/delegates/ShadowActivityDelegate.kt
+++ b/projects/sdk/core/loader/src/main/kotlin/com/tencent/shadow/core/loader/delegates/ShadowActivityDelegate.kt
@@ -26,6 +26,7 @@ import android.content.Intent
 import android.content.pm.ActivityInfo
 import android.content.res.Configuration
 import android.content.res.Resources
+import android.os.Build
 import android.os.Bundle
 import android.view.LayoutInflater
 import android.view.WindowManager
@@ -39,9 +40,7 @@ import com.tencent.shadow.core.loader.managers.ComponentManager.Companion.CM_CLA
 import com.tencent.shadow.core.loader.managers.ComponentManager.Companion.CM_EXTRAS_BUNDLE_KEY
 import com.tencent.shadow.core.loader.managers.ComponentManager.Companion.CM_LOADER_BUNDLE_KEY
 import com.tencent.shadow.core.loader.managers.ComponentManager.Companion.CM_PART_KEY
-import com.tencent.shadow.core.runtime.MixResources
-import com.tencent.shadow.core.runtime.PluginActivity
-import com.tencent.shadow.core.runtime.ShadowLayoutInflater
+import com.tencent.shadow.core.runtime.*
 import com.tencent.shadow.core.runtime.container.HostActivityDelegate
 import com.tencent.shadow.core.runtime.container.HostActivityDelegator
 
@@ -135,12 +134,18 @@ class ShadowActivityDelegate(private val mDI: DI) : GeneratedShadowActivityDeleg
 
             //Activity.onCreate调用之前应该先收到onWindowAttributesChanged。
             if (mCallOnWindowAttributesChanged) {
-                pluginActivity.onWindowAttributesChanged(mBeforeOnCreateOnWindowAttributesChangedCalledParams)
+                pluginActivity.onWindowAttributesChanged(
+                    mBeforeOnCreateOnWindowAttributesChangedCalledParams
+                )
                 mBeforeOnCreateOnWindowAttributesChangedCalledParams = null
             }
 
-            val pluginSavedInstanceState: Bundle? = savedInstanceState?.getBundle(PLUGIN_OUT_STATE_KEY)
+            val pluginSavedInstanceState: Bundle? =
+                savedInstanceState?.getBundle(PLUGIN_OUT_STATE_KEY)
             pluginSavedInstanceState?.classLoader = mPluginClassLoader
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                notifyPluginActivityPreCreated(pluginActivity, pluginSavedInstanceState)
+            }
             pluginActivity.onCreate(pluginSavedInstanceState)
             mPluginActivityCreated = true
         } catch (e: Exception) {
@@ -262,4 +267,13 @@ class ShadowActivityDelegate(private val mDI: DI) : GeneratedShadowActivityDeleg
         mHostActivityDelegator.superRecreate()
     }
 
+    private fun notifyPluginActivityPreCreated(
+        pluginActivity: ShadowActivity,
+        pluginSavedInstanceState: Bundle?
+    ) {
+        ShadowActivityLifecycleCallbacks.Holder.notifyPluginActivityPreCreated(
+            pluginActivity,
+            pluginSavedInstanceState
+        )
+    }
 }

--- a/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/PluginActivity.java
+++ b/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/PluginActivity.java
@@ -46,6 +46,18 @@ public abstract class PluginActivity extends GeneratedPluginActivity {
 
     ComponentName mCallingActivity;
 
+    private ShadowActivityLifecycleCallbacks.Holder lifecycleCallbacksHolder;
+
+    public void registerActivityLifecycleCallbacks(
+            ShadowActivityLifecycleCallbacks callback) {
+        lifecycleCallbacksHolder.registerActivityLifecycleCallbacks(this, callback);
+    }
+
+    public void unregisterActivityLifecycleCallbacks(
+            ShadowActivityLifecycleCallbacks callback) {
+        lifecycleCallbacksHolder.unregisterActivityLifecycleCallbacks(callback);
+    }
+
     public final void setHostContextAsBase(Context context) {
         attachBaseContext(context);
     }
@@ -53,6 +65,7 @@ public abstract class PluginActivity extends GeneratedPluginActivity {
     public void setHostActivityDelegator(HostActivityDelegator delegator) {
         super.hostActivityDelegator = delegator;
         hostActivityDelegator = delegator;
+        lifecycleCallbacksHolder = new ShadowActivityLifecycleCallbacks.Holder(delegator);
     }
 
     public void setPluginApplication(ShadowApplication pluginApplication) {

--- a/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/ShadowActivity.java
+++ b/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/ShadowActivity.java
@@ -18,7 +18,6 @@
 
 package com.tencent.shadow.core.runtime;
 
-import android.app.Application;
 import android.content.ComponentName;
 import android.content.Intent;
 import android.content.IntentSender;
@@ -27,9 +26,6 @@ import android.os.Bundle;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
-
-import java.util.HashMap;
-import java.util.Map;
 
 public class ShadowActivity extends PluginActivity {
 
@@ -116,32 +112,6 @@ public class ShadowActivity extends PluginActivity {
             throw new IllegalArgumentException("ID does not reference a View inside this Activity");
         }
         return view;
-    }
-
-    final private Map<ShadowActivityLifecycleCallbacks,
-            Application.ActivityLifecycleCallbacks>
-            mActivityLifecycleCallbacksMap = new HashMap<>();
-
-    public void registerActivityLifecycleCallbacks(
-            ShadowActivityLifecycleCallbacks callback) {
-        synchronized (mActivityLifecycleCallbacksMap) {
-            final ShadowActivityLifecycleCallbacks.Wrapper wrapper
-                    = new ShadowActivityLifecycleCallbacks.Wrapper(callback, this);
-            mActivityLifecycleCallbacksMap.put(callback, wrapper);
-            hostActivityDelegator.registerActivityLifecycleCallbacks(wrapper);
-        }
-    }
-
-    public void unregisterActivityLifecycleCallbacks(
-            ShadowActivityLifecycleCallbacks callback) {
-        synchronized (mActivityLifecycleCallbacksMap) {
-            final Application.ActivityLifecycleCallbacks activityLifecycleCallbacks
-                    = mActivityLifecycleCallbacksMap.get(callback);
-            if (activityLifecycleCallbacks != null) {
-                hostActivityDelegator.unregisterActivityLifecycleCallbacks(activityLifecycleCallbacks);
-                mActivityLifecycleCallbacksMap.remove(callback);
-            }
-        }
     }
 
     @Override

--- a/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/ShadowActivityLifecycleCallbacks.java
+++ b/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/ShadowActivityLifecycleCallbacks.java
@@ -22,7 +22,13 @@ import android.app.Activity;
 import android.app.Application;
 import android.os.Bundle;
 
+import com.tencent.shadow.core.runtime.container.HostActivityDelegator;
 import com.tencent.shadow.core.runtime.container.PluginContainerActivity;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
 
 public interface ShadowActivityLifecycleCallbacks {
 
@@ -162,7 +168,10 @@ public interface ShadowActivityLifecycleCallbacks {
 
         @Override
         public void onActivityPreCreated(Activity activity, Bundle savedInstanceState) {
-            final ShadowActivity pluginActivity = getPluginActivity(activity);
+            //此时PluginActivity尚未构造。改由onPluginActivityPreCreated通知。
+        }
+
+        public void onPluginActivityPreCreated(ShadowActivity pluginActivity, Bundle savedInstanceState) {
             if (checkOwnerActivity(pluginActivity)) {
                 try {
                     shadowActivityLifecycleCallbacks.onActivityPreCreated(pluginActivity, savedInstanceState);
@@ -341,6 +350,94 @@ public interface ShadowActivityLifecycleCallbacks {
                 return activity.mPluginApplication == runtimeObject;
             } else {
                 return activity == runtimeObject;
+            }
+        }
+    }
+
+    /**
+     * ShadowActivity和ShadowApplication共用的ActivityLifecycleCallbacks逻辑。
+     * 用一个Map将ActivityLifecycleCallbacks映射成ShadowActivityLifecycleCallbacks。
+     * 同时，持有所有的有注册ActivityLifecycleCallbacks的Holder，
+     * 提供通知所有callbacks onPluginActivityPreCreated事件的方法。
+     */
+    class Holder {
+        final private static Set<Holder> sAllHolders = new HashSet<>();
+
+        final private HostActivityDelegator delegator;
+        final private Application application;
+
+        final private Map<ShadowActivityLifecycleCallbacks,
+                Wrapper>
+                mActivityLifecycleCallbacksMap = new HashMap<>();
+
+        public Holder(Object hostActivityDelegatorOrApplication) {
+            if (hostActivityDelegatorOrApplication instanceof HostActivityDelegator) {
+                this.delegator = (HostActivityDelegator) hostActivityDelegatorOrApplication;
+                this.application = null;
+            } else {
+                this.delegator = null;
+                this.application = (Application) hostActivityDelegatorOrApplication;
+            }
+        }
+
+        private void register(Application.ActivityLifecycleCallbacks callbacks) {
+            if (delegator != null) {
+                delegator.registerActivityLifecycleCallbacks(callbacks);
+            } else {
+                assert application != null;
+                application.registerActivityLifecycleCallbacks(callbacks);
+            }
+        }
+
+        private void unregister(Application.ActivityLifecycleCallbacks callbacks) {
+            if (delegator != null) {
+                delegator.unregisterActivityLifecycleCallbacks(callbacks);
+            } else {
+                assert application != null;
+                application.unregisterActivityLifecycleCallbacks(callbacks);
+            }
+        }
+
+        public void registerActivityLifecycleCallbacks(
+                Object caller,
+                ShadowActivityLifecycleCallbacks callback) {
+            synchronized (sAllHolders) {
+                final ShadowActivityLifecycleCallbacks.Wrapper wrapper
+                        = new ShadowActivityLifecycleCallbacks.Wrapper(callback, caller);
+                mActivityLifecycleCallbacksMap.put(callback, wrapper);
+                register(wrapper);
+                sAllHolders.add(this);
+            }
+        }
+
+        public void unregisterActivityLifecycleCallbacks(
+                ShadowActivityLifecycleCallbacks callback) {
+            synchronized (sAllHolders) {
+                final Application.ActivityLifecycleCallbacks activityLifecycleCallbacks
+                        = mActivityLifecycleCallbacksMap.get(callback);
+                if (activityLifecycleCallbacks != null) {
+                    unregister(activityLifecycleCallbacks);
+                    mActivityLifecycleCallbacksMap.remove(callback);
+                }
+                if (mActivityLifecycleCallbacksMap.isEmpty()) {
+                    sAllHolders.remove(this);
+                }
+            }
+        }
+
+        public static void notifyPluginActivityPreCreated(ShadowActivity pluginActivity, Bundle savedInstanceState) {
+            synchronized (sAllHolders) {
+                for (Holder holder : sAllHolders) {
+                    holder.onPluginActivityPreCreated(pluginActivity, savedInstanceState);
+                }
+            }
+        }
+
+        private void onPluginActivityPreCreated(ShadowActivity pluginActivity, Bundle savedInstanceState) {
+            synchronized (sAllHolders) {
+                for (ShadowActivityLifecycleCallbacks.Wrapper wrapper : mActivityLifecycleCallbacksMap.values()) {
+                    wrapper.onPluginActivityPreCreated(pluginActivity, savedInstanceState);
+                }
             }
         }
     }

--- a/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/ShadowApplication.java
+++ b/projects/sdk/core/runtime/src/main/java/com/tencent/shadow/core/runtime/ShadowApplication.java
@@ -28,7 +28,6 @@ import android.content.IntentFilter;
 import android.content.res.Configuration;
 import android.os.Build;
 
-import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
@@ -50,15 +49,23 @@ public class ShadowApplication extends ShadowContext {
         return this;
     }
 
-    final private Map<ShadowActivityLifecycleCallbacks,
-            Application.ActivityLifecycleCallbacks>
-            mActivityLifecycleCallbacksMap = new HashMap<>();
+    private ShadowActivityLifecycleCallbacks.Holder lifecycleCallbacksHolder;
+
+    public void registerActivityLifecycleCallbacks(
+            ShadowActivityLifecycleCallbacks callback) {
+        lifecycleCallbacksHolder.registerActivityLifecycleCallbacks(this, callback);
+    }
+
+    public void unregisterActivityLifecycleCallbacks(
+            ShadowActivityLifecycleCallbacks callback) {
+        lifecycleCallbacksHolder.unregisterActivityLifecycleCallbacks(callback);
+    }
 
     public void onCreate() {
 
         isCallOnCreate = true;
 
-        for (Map.Entry<String, List<String>> entry: mBroadcasts.entrySet()){
+        for (Map.Entry<String, List<String>> entry : mBroadcasts.entrySet()) {
             try {
                 Class<?> clazz = mPluginClassLoader.loadClass(entry.getKey());
                 BroadcastReceiver receiver = ((BroadcastReceiver) clazz.newInstance());
@@ -122,29 +129,6 @@ public class ShadowApplication extends ShadowContext {
         mHostApplication.unregisterComponentCallbacks(callback);
     }
 
-
-    public void registerActivityLifecycleCallbacks(ShadowActivityLifecycleCallbacks callback) {
-        synchronized (mActivityLifecycleCallbacksMap) {
-            final ShadowActivityLifecycleCallbacks.Wrapper wrapper
-                    = new ShadowActivityLifecycleCallbacks.Wrapper(callback, this);
-            mActivityLifecycleCallbacksMap.put(callback, wrapper);
-            mHostApplication.registerActivityLifecycleCallbacks(wrapper);
-        }
-    }
-
-
-    public void unregisterActivityLifecycleCallbacks(ShadowActivityLifecycleCallbacks callback) {
-        synchronized (mActivityLifecycleCallbacksMap) {
-            final Application.ActivityLifecycleCallbacks activityLifecycleCallbacks
-                    = mActivityLifecycleCallbacksMap.get(callback);
-            if (activityLifecycleCallbacks != null) {
-                mHostApplication.unregisterActivityLifecycleCallbacks(activityLifecycleCallbacks);
-                mActivityLifecycleCallbacksMap.remove(callback);
-            }
-        }
-    }
-
-
     @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
     public void registerOnProvideAssistDataListener(Application.OnProvideAssistDataListener callback) {
         mHostApplication.registerOnProvideAssistDataListener(callback);
@@ -159,6 +143,8 @@ public class ShadowApplication extends ShadowContext {
     public void setHostApplicationContextAsBase(Context hostAppContext) {
         super.attachBaseContext(hostAppContext);
         mHostApplication = (Application) hostAppContext;
+        lifecycleCallbacksHolder
+                = new ShadowActivityLifecycleCallbacks.Holder(mHostApplication);
     }
 
     public void setBroadcasts(Map<String, List<String>> broadcast){

--- a/projects/test/dynamic/host/test-dynamic-host/src/androidTest/java/com/tencent/shadow/test/cases/plugin_main/ActivityLifecycleCallbacksTest.java
+++ b/projects/test/dynamic/host/test-dynamic-host/src/androidTest/java/com/tencent/shadow/test/cases/plugin_main/ActivityLifecycleCallbacksTest.java
@@ -1,0 +1,74 @@
+package com.tencent.shadow.test.cases.plugin_main;
+
+import android.content.Intent;
+import android.os.Build;
+
+import androidx.test.core.app.ApplicationProvider;
+import androidx.test.espresso.Espresso;
+import androidx.test.espresso.action.ViewActions;
+import androidx.test.espresso.matcher.ViewMatchers;
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+
+import org.hamcrest.Matchers;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@RunWith(AndroidJUnit4.class)
+public class ActivityLifecycleCallbacksTest extends PluginMainAppTest {
+
+    @Override
+    protected Intent getLaunchIntent() {
+        Intent pluginIntent = new Intent();
+        String packageName = ApplicationProvider.getApplicationContext().getPackageName();
+        pluginIntent.setClassName(
+                packageName,
+                "com.tencent.shadow.test.plugin.general_cases.lib.usecases.application.ActivityLifecycleCallbacksTestActivity"
+        );
+        return pluginIntent;
+    }
+
+
+    @Test
+    public void testRecreateRecord() {
+        Espresso.onView(ViewMatchers.withTagValue(Matchers.<Object>is("recreate")))
+                .perform(ViewActions.click());
+
+        String api29 = "[onActivityPreCreated, " +
+                "onActivityCreated, " +
+                "onActivityPostCreated, " +
+                "onActivityPreStarted, " +
+                "onActivityStarted, " +
+                "onActivityPostStarted, " +
+                "onActivityPreResumed, " +
+                "onActivityResumed, " +
+                "onActivityPostResumed, " +
+                "onActivityPrePaused, " +
+                "onActivityPaused, " +
+                "onActivityPostPaused, " +
+                "onActivityPreStopped, " +
+                "onActivityStopped, " +
+                "onActivityPostStopped, " +
+                "onActivityPreSaveInstanceState, " +
+                "onActivitySaveInstanceState, " +
+                "onActivityPostSaveInstanceState, " +
+                "onActivityPreDestroyed, " +
+                "onActivityDestroyed, " +
+                "onActivityPostDestroyed, " +
+                "onActivityPreCreated, " +
+                "onActivityCreated]";
+
+        String api28 = "[onActivityCreated, " +
+                "onActivityStarted, " +
+                "onActivityResumed, " +
+                "onActivityPaused, " +
+                "onActivityStopped, " +
+                "onActivitySaveInstanceState, " +
+                "onActivityDestroyed, " +
+                "onActivityCreated]";
+
+        String expect = Build.VERSION.SDK_INT < Build.VERSION_CODES.Q ? api28 : api29;
+
+        matchTextWithViewTag("ActivityCreateTest", expect);
+    }
+
+}

--- a/projects/test/plugin/general-cases/general-cases-app/src/androidTest/java/com/tencent/shadow/test/plugin/general_cases/app/ActivityLifecycleCallbacksTest.java
+++ b/projects/test/plugin/general-cases/general-cases-app/src/androidTest/java/com/tencent/shadow/test/plugin/general_cases/app/ActivityLifecycleCallbacksTest.java
@@ -1,0 +1,75 @@
+package com.tencent.shadow.test.plugin.general_cases.app;
+
+import android.content.Intent;
+import android.os.Build;
+
+import androidx.test.core.app.ActivityScenario;
+import androidx.test.core.app.ApplicationProvider;
+import androidx.test.espresso.Espresso;
+import androidx.test.espresso.action.ViewActions;
+import androidx.test.espresso.matcher.ViewMatchers;
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+
+import org.hamcrest.Matchers;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@RunWith(AndroidJUnit4.class)
+public class ActivityLifecycleCallbacksTest extends NormalAppTest {
+
+    @Before
+    public void launchActivity() {
+        Intent intent = new Intent();
+        String packageName = ApplicationProvider.getApplicationContext().getPackageName();
+        intent.setClassName(
+                packageName,
+                "com.tencent.shadow.test.plugin.general_cases.lib.usecases.application.ActivityLifecycleCallbacksTestActivity"
+        );
+        ActivityScenario.launch(intent);
+    }
+
+    @Test
+    public void testRecreateRecord() {
+        Espresso.onView(ViewMatchers.withTagValue(Matchers.<Object>is("recreate")))
+                .perform(ViewActions.click());
+
+        String api29 = "[onActivityPreCreated, " +
+                "onActivityCreated, " +
+                "onActivityPostCreated, " +
+                "onActivityPreStarted, " +
+                "onActivityStarted, " +
+                "onActivityPostStarted, " +
+                "onActivityPreResumed, " +
+                "onActivityResumed, " +
+                "onActivityPostResumed, " +
+                "onActivityPrePaused, " +
+                "onActivityPaused, " +
+                "onActivityPostPaused, " +
+                "onActivityPreStopped, " +
+                "onActivityStopped, " +
+                "onActivityPostStopped, " +
+                "onActivityPreSaveInstanceState, " +
+                "onActivitySaveInstanceState, " +
+                "onActivityPostSaveInstanceState, " +
+                "onActivityPreDestroyed, " +
+                "onActivityDestroyed, " +
+                "onActivityPostDestroyed, " +
+                "onActivityPreCreated, " +
+                "onActivityCreated]";
+
+        String api28 = "[onActivityCreated, " +
+                "onActivityStarted, " +
+                "onActivityResumed, " +
+                "onActivityPaused, " +
+                "onActivityStopped, " +
+                "onActivitySaveInstanceState, " +
+                "onActivityDestroyed, " +
+                "onActivityCreated]";
+
+        String expect = Build.VERSION.SDK_INT < Build.VERSION_CODES.Q ? api28 : api29;
+
+        matchTextWithViewTag("ActivityCreateTest", expect);
+    }
+
+}

--- a/projects/test/plugin/general-cases/general-cases-lib/src/main/AndroidManifest.xml
+++ b/projects/test/plugin/general-cases/general-cases-lib/src/main/AndroidManifest.xml
@@ -85,6 +85,7 @@
         <activity android:name=".usecases.context.RegisterNullReceiverActivity" />
         <activity android:name=".usecases.dialog.TestAlertDialogActivity" />
         <activity android:name=".usecases.instrumentation.TestInstrumentationActivity" />
+        <activity android:name=".usecases.application.ActivityLifecycleCallbacksTestActivity" />
 
         <provider
             android:authorities="com.tencent.shadow.provider.test"

--- a/projects/test/plugin/general-cases/general-cases-lib/src/main/java/com/tencent/shadow/test/plugin/general_cases/lib/gallery/TestApplication.java
+++ b/projects/test/plugin/general-cases/general-cases-lib/src/main/java/com/tencent/shadow/test/plugin/general_cases/lib/gallery/TestApplication.java
@@ -18,7 +18,14 @@
 
 package com.tencent.shadow.test.plugin.general_cases.lib.gallery;
 
+import android.app.Activity;
 import android.app.Application;
+import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.LinkedList;
+import java.util.List;
 
 public class TestApplication extends Application {
 
@@ -26,14 +33,183 @@ public class TestApplication extends Application {
 
     public boolean isOnCreate;
 
+    final private TestActivityLifecycleCallbacks alc = new TestActivityLifecycleCallbacks(
+            "com.tencent.shadow.test.plugin.general_cases.lib.usecases.application.ActivityLifecycleCallbacksTestActivity");
+
     @Override
     public void onCreate() {
         sInstence = this;
         isOnCreate = true;
         super.onCreate();
+
+        registerActivityLifecycleCallbacks(alc);
     }
 
     public static TestApplication getInstance() {
         return sInstence;
+    }
+
+    public List<String> getTestActivityLifecycleCallbacksRecord() {
+        return alc.recordList;
+    }
+}
+
+class TestActivityLifecycleCallbacks implements Application.ActivityLifecycleCallbacks {
+    final private String targetActivityName;
+    final List<String> recordList = new LinkedList<>();
+
+    TestActivityLifecycleCallbacks(String targetActivityName) {
+        this.targetActivityName = targetActivityName;
+    }
+
+    private boolean isTargetActivity(Activity activity) {
+        return activity.getClass().getName().equals(targetActivityName);
+    }
+
+    @Override
+    public void onActivityPreCreated(@NonNull Activity activity, @Nullable Bundle savedInstanceState) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPreCreated");
+        }
+    }
+
+    @Override
+    public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle savedInstanceState) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityCreated");
+        }
+    }
+
+    @Override
+    public void onActivityPostCreated(@NonNull Activity activity, @Nullable Bundle savedInstanceState) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPostCreated");
+        }
+    }
+
+    @Override
+    public void onActivityPreStarted(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPreStarted");
+        }
+    }
+
+    @Override
+    public void onActivityStarted(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityStarted");
+        }
+    }
+
+    @Override
+    public void onActivityPostStarted(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPostStarted");
+        }
+    }
+
+    @Override
+    public void onActivityPreResumed(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPreResumed");
+        }
+    }
+
+    @Override
+    public void onActivityResumed(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityResumed");
+        }
+    }
+
+    @Override
+    public void onActivityPostResumed(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPostResumed");
+        }
+    }
+
+    @Override
+    public void onActivityPrePaused(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPrePaused");
+        }
+    }
+
+    @Override
+    public void onActivityPaused(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPaused");
+        }
+    }
+
+    @Override
+    public void onActivityPostPaused(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPostPaused");
+        }
+    }
+
+    @Override
+    public void onActivityPreStopped(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPreStopped");
+        }
+    }
+
+    @Override
+    public void onActivityStopped(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityStopped");
+        }
+    }
+
+    @Override
+    public void onActivityPostStopped(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPostStopped");
+        }
+    }
+
+    @Override
+    public void onActivityPreSaveInstanceState(@NonNull Activity activity, @NonNull Bundle outState) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPreSaveInstanceState");
+        }
+    }
+
+    @Override
+    public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle outState) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivitySaveInstanceState");
+        }
+    }
+
+    @Override
+    public void onActivityPostSaveInstanceState(@NonNull Activity activity, @NonNull Bundle outState) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPostSaveInstanceState");
+        }
+    }
+
+    @Override
+    public void onActivityPreDestroyed(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPreDestroyed");
+        }
+    }
+
+    @Override
+    public void onActivityDestroyed(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityDestroyed");
+        }
+    }
+
+    @Override
+    public void onActivityPostDestroyed(@NonNull Activity activity) {
+        if (isTargetActivity(activity)) {
+            recordList.add("onActivityPostDestroyed");
+        }
     }
 }

--- a/projects/test/plugin/general-cases/general-cases-lib/src/main/java/com/tencent/shadow/test/plugin/general_cases/lib/usecases/application/ActivityLifecycleCallbacksTestActivity.java
+++ b/projects/test/plugin/general-cases/general-cases-lib/src/main/java/com/tencent/shadow/test/plugin/general_cases/lib/usecases/application/ActivityLifecycleCallbacksTestActivity.java
@@ -1,0 +1,54 @@
+package com.tencent.shadow.test.plugin.general_cases.lib.usecases.application;
+
+import android.app.Activity;
+import android.os.Bundle;
+import android.support.annotation.Nullable;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.Button;
+
+import com.tencent.shadow.test.plugin.general_cases.lib.gallery.TestApplication;
+import com.tencent.shadow.test.plugin.general_cases.lib.gallery.util.UiUtil;
+
+import java.util.List;
+
+/**
+ * 在
+ * com.tencent.shadow.test.plugin.general_cases.lib.gallery.TestApplication
+ * 中注册一个ActivityLifecycleCallbacks，专门监听
+ * com.tencent.shadow.test.plugin.general_cases.lib.usecases.application.TestApplicationActivity
+ * 的生命周期。
+ * 然后用这个Activity打印出监听记录进行测试。
+ */
+public class ActivityLifecycleCallbacksTestActivity extends Activity {
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        ViewGroup viewGroup = UiUtil.setActivityContentView(this);
+
+        List<String> record = TestApplication.getInstance().getTestActivityLifecycleCallbacksRecord();
+
+
+        viewGroup.addView(
+                UiUtil.makeItem(
+                        this,
+                        "ActivityCreateTest",
+                        "ActivityCreateTest",
+                        record.toString()
+                )
+        );
+
+        Button recreateButton = new Button(this);
+        recreateButton.setText("recreate");
+        recreateButton.setTag("recreate");
+        recreateButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                ActivityLifecycleCallbacksTestActivity.this.recreate();
+            }
+        });
+        viewGroup.addView(recreateButton);
+    }
+
+}

